File: python_solutions/678.py
class Solution:
    def checkValidString(self, s: str) -> bool:
        star_stack = []
        open_stack = []
        for idx,c in enumerate(s):
            if c == '(':
                open_stack.append(idx)
            elif c == '*':
                star_stack.append(idx)
            else:
                if len(open_stack) > 0:
                    open_stack.pop()
                elif len(star_stack) > 0:
                    star_stack.pop()
                else:
                    return False
        while len(open_stack) > 0:
            if len(star_stack) == 0:
                return False
            elif star_stack.pop() < open_stack.pop():
                return False
        return True
            
            
        

File: rust_solutions/678.rs
impl Solution {
    pub fn check_valid_string(s: String) -> bool {
        let mut open = Vec::new();
        let mut star = Vec::new();
        for (idx,c) in s.chars().enumerate() {
            match c {
                '(' => {open.push(idx);},
                '*' => {star.push(idx);},
                _ => {
                    if open.len() > 0{
                        open.pop();
                    }
                    else if star.len() > 0 {
                        star.pop();
                    }
                    else{
                        return false;
                    }
                }
            }
        }
        
        while open.len() > 0 {
            if star.len() == 0 {
                return false;
            }
            match (open.pop() , star.pop()) {
                (Some(x),Some(y)) => {if x > y {return false;}},
                (Some(x),None) => {return false;},
                _ => {break ;}
            }
        }
        
        true
    }
}
